v2/protocol/pubsub/protocol.go

```diff
@@ -0,0 +1,247 @@
+package pubsub
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub/internal"
+)
+
+const (
+	ProtocolName = "Pub/Sub"
+)
+
+type subscriptionWithTopic struct {
+	topicID        string
+	subscriptionID string
+}
+
+// Protocol acts as both a pubsub topic and a pubsub subscription .
+type Protocol struct {
+	transformers binding.TransformerFactories
+
+	// PubSub
+
+	// ReceiveSettings is used to configure Pubsub pull subscription.
+	ReceiveSettings *pubsub.ReceiveSettings
+
+	// AllowCreateTopic controls if the transport can create a topic if it does
+	// not exist.
+	AllowCreateTopic bool
+
+	// AllowCreateSubscription controls if the transport can create a
+	// subscription if it does not exist.
+	AllowCreateSubscription bool
+
+	projectID      string
+	topicID        string
+	subscriptionID string
+
+	gccMux sync.Mutex
+
+	subscriptions []subscriptionWithTopic
+	client        *pubsub.Client
+
+	connectionsBySubscription map[string]*internal.Connection
+	connectionsByTopic        map[string]*internal.Connection
+
+	incoming chan pubsub.Message
+}
+
+// New creates a new pubsub transport.
+func New(ctx context.Context, opts ...Option) (*Protocol, error) {
+	t := &Protocol{}
+	t.incoming = make(chan pubsub.Message)
+	if err := t.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if t.client == nil {
+		// Auth to pubsub.
+		client, err := pubsub.NewClient(ctx, t.projectID)
+		if err != nil {
+			return nil, err
+		}
+		// Success.
+		t.client = client
+	}
+
+	if t.connectionsBySubscription == nil {
+		t.connectionsBySubscription = make(map[string]*internal.Connection, 0)
+	}
+
+	if t.connectionsByTopic == nil {
+		t.connectionsByTopic = make(map[string]*internal.Connection, 0)
+	}
+	return t, nil
+}
+
+func (t *Protocol) applyOptions(opts ...Option) error {
+	for _, fn := range opts {
+		if err := fn(t); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Send implements Sender.Send
+func (t *Protocol) Send(ctx context.Context, in binding.Message) error {
+	topic := cecontext.TopicFrom(ctx)
```

Comment:
`Sender` & `Requester` contract requires that `Send` invokes `Finish` on the input message

Reply:
done